libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/StackLimitEvader.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+// Walks the call graph using a Depth-First-Search assigning memory offsets to variables.
+// - The leaves of the call graph will get the lowest offsets, increasing towards the root.
+// - ``nextAvailableSlot`` maps a function to the next available slot that can be used by another
```

Comment:
Terminology: Slot and offset is the same?

Reply:
Maybe it would be easier to understand if we just count the number of slots needed by the function and its callees?

Reply:
If I see it correctly, `nextSlot` can be renamed `slotsNeeded` and it would still work, right? If that is true, I would prefer to avoid the term "next" since it always sounds like it has to refer to something or needs a certain ordering of elements.

Reply:
Rephrased and renamed everything a bit.